pkg/indexer/atlasdatabaseuserprojects.go

```diff
@@ -0,0 +1,48 @@
+package indexer
+
+import (
+	"go.uber.org/zap"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+)
+
+const (
+	AtlasDatabaseUserByProjectsIndex = "atlasdatabaseuser.spec.project"
+)
+
+type AtlasDatabaseUserByProjectsIndexer struct {
+	logger *zap.SugaredLogger
+}
+
+func NewAtlasDatabaseUserByProjectsIndexer(logger *zap.Logger) *AtlasDatabaseUserByProjectsIndexer {
+	return &AtlasDatabaseUserByProjectsIndexer{
+		logger: logger.Named(AtlasDatabaseUserByProjectsIndex).Sugar(),
+	}
+}
+
+func (*AtlasDatabaseUserByProjectsIndexer) Object() client.Object {
+	return &akov2.AtlasDatabaseUser{}
+}
+
+func (*AtlasDatabaseUserByProjectsIndexer) Name() string {
+	return AtlasDatabaseUserByProjectsIndex
+}
+
+func (a *AtlasDatabaseUserByProjectsIndexer) Keys(object client.Object) []string {
+	user, ok := object.(*akov2.AtlasDatabaseUser)
+	if !ok {
+		a.logger.Errorf("expected *akov2.AtlasDatabaseUser but got %T", object)
+		return nil
+	}
+
+	if user.Spec.Project != nil && user.Spec.Project.Name != "" {
+		return []string{user.Spec.Project.GetObject(user.Namespace).String()}
+	}
+
+	if user.Spec.AtlasRef != nil && user.Spec.AtlasRef.ID != "" {
+		return []string{user.Spec.AtlasRef.ID}
+	}
```

Comment:
This is mixing two independent indexes with different keys: 1. using AKO project names as keys pointing to database users and 2. one using Atlas Project IDs pointing to database users. For the latter the index name `atlasdatabaseuser.spec.project` is not correct.

Please create a dedicated index with a correct name for atlas IDs.

If we want to replace boilerplate we could:
a) rename this index as `atlasdatabaseuser.spec.project` is not correct
b) remove the precedence of project name keys over atlas project IDs pointing to database users. That precedence is not obvious. If it is necessary it should be documented and tested.
c) Prefix each separate key, i.e. using `spec.project/fooNamespace/fooProject` and `spec.atlasRef/123456..`

Having said that I still very much prefer a separate index as it is one.